### PR TITLE
Add com.vba_m.visualboyadvance-m

### PR DESCRIPTION
**Pull request by ZachBacon**
### Please confirm your submission meets all the criteria

- [x] I have read the [App Requirements][reqs] and [App Maintenance][maint] pages.
- [x] My pull request follows the instructions at [App Submission][submission].
- [x] I have [built][build] and tested the submission locally.
- [x] I am using only the minimal set of permissions. *(If not, please explain each non-standard permission.)*
- [x] All assets referenced in the manifest are redistributable by any party.  If not, the unredistributable parts are using an extra-data source type.
- [x] I am an upstream contributor to the project. If not, I contacted upstream developers about submitting their software to Flathub. **Link:** https://github.com/visualboyadvance-m/visualboyadvance-m https://vba-m.com
- [x] I own the domain used in the application ID or the domain has a policy for delegating subdomains (e.g. GitHub, SourceForge).
- [x] Any additional patches or files have been submitted to the upstream projects concerned. *(If not, explain why.)*

[reqs]: https://docs.flathub.org/docs/for-app-authors/requirements
[maint]: https://docs.flathub.org/docs/for-app-authors/maintanance
[submission]: https://docs.flathub.org/docs/for-app-authors/submission
[build]: https://docs.flathub.org/docs/for-app-authors/submission/#before-submission

### CODE REVIEW

File: com.vbam.visualboyadvancem.yaml [hunk 1]
```diff
@@ -0,0 +1,119 @@
+app-id: com.vbam.visualboyadvancem
+runtime: org.freedesktop.Platform
+runtime-version: '22.08'
```
Comment (hfiguiere) [inline com.vbam.visualboyadvancem.yaml:3]:
```suggestion
runtime-version: '23.08'
```

it's the current version

File: com.vbam.visualboyadvancem.yaml [hunk 2]
```diff
@@ -0,0 +1,119 @@
...
+      - --enable-avfilter
+      - --enable-libxml2
+      - --enable-swscale
+
+      - --enable-libxcb
+
+      - --enable-protocol=file
+    sources:
+      - type: archive
+        url: https://ffmpeg.org/releases/ffmpeg-6.1.tar.xz
```
Comment (hfiguiere) [inline com.vbam.visualboyadvancem.yaml:106]: 6.0 is in the runtime. isn't it enough?

File: com.vbam.visualboyadvancem.yaml [hunk 3]
```diff
@@ -0,0 +1,119 @@
...
+rename-icon: visualboyadvance-m
+rename-desktop-file: visualboyadvance-m.desktop
+finish-args:
+  - --device=all
+  - --share=network
+  - --share=ipc
+  - --socket=fallback-x11
+  - --socket=wayland
+  - --socket=pulseaudio
+  - --filesystem=host:ro
```
Comment (hfiguiere) [inline com.vbam.visualboyadvancem.yaml:15]: why host filesystem access?

File: com.vbam.visualboyadvancem.yaml [hunk 4]
```diff
@@ -0,0 +1,119 @@
+app-id: com.vbam.visualboyadvancem
```
Comment (hfiguiere) [inline com.vbam.visualboyadvancem.yaml:1]: `vbam.com` doesn't look related to this. Is this a domain controlled by the upstream developers?

Comment (ZachBacon) [reply com.vbam.visualboyadvancem.yaml:1]: Sadly it's vba-m.com but when I tried to build it, flatpak-builder was complaining about the dash for some reason.

As for ffmpeg. I opted for a more minimalistic build over the full bodied compile the SDK provides.

File: com.vbam.visualboyadvancem.yaml (hunk 3 above)
Comment (ZachBacon) [reply com.vbam.visualboyadvancem.yaml:15]: How is is someone going to load a rom for a gba emulator?

File: com.vbam.visualboyadvancem.yaml (hunk 4 above)
Comment (hfiguiere) [reply com.vbam.visualboyadvancem.yaml:1]: because `-` is to be replaced with `_`. One doesn't simply take a domain from someone else.

there is no point in building ffmpeg. It's just a waste of space.

Comment (ZachBacon) [reply com.vbam.visualboyadvancem.yaml:1]: specifically vba-m.com being the domain, but flatpak-builder doesn't like that. 

``error: 'com.vba-m.visualboyadvance-m' is not a valid application name: Only last name segment can contain -``
``error: 'com.vba-m.visualboyadvancem' is not a valid application name: Only last name segment can contain -``

Comment (ZachBacon) [reply com.vbam.visualboyadvancem.yaml:1]: Also I'm one of the upstream developers.

Comment (hfiguiere) [reply com.vbam.visualboyadvancem.yaml:1]: as I just said, replace `-` with `_`.

https://docs.flatpak.org/en/latest/conventions.html#application-ids

Comment (ZachBacon) [reply com.vbam.visualboyadvancem.yaml:1]: > there is no point in building ffmpeg. It's just a waste of space.

Fine, I get it. Also, apologies. didn't see the previous comment. I'll resolve those issues in a moment.

File: com.vbam.visualboyadvancem.yaml (hunk 3 above)
Comment (ZachBacon) [reply com.vbam.visualboyadvancem.yaml:15]: though that might need to be rw, mainly for the battery saves. Usually that gets dumped beside the rom. Is that going to be a potential problem?

File: com.vbam.visualboyadvancem.yaml (hunk 4 above)
Comment (ZachBacon) [reply com.vbam.visualboyadvancem.yaml]: Alright, new commit should resolve most of these current issues.

**Comment by hfiguiere** [comment]
bot, build com.vba_m.visualboyadvance-m

**Comment by ZachBacon** [comment]
Just saw the other errors in the build. I'm going to fix them in a moment. I already made a commit in vba-m's own git regarding the screenshots in the metainfo. I'll grab the patch from that commit and bring them over.

**Comment by ZachBacon** [comment]
Alright, that should fix the appstream validation. Ran it through locally using appstream-util, and backported the patch for the flatpak. It was needing an update anyways.

File: com.vba_m.visualboyadvance-m.yaml [hunk 5]
```diff
@@ -0,0 +1,76 @@
...
+  - /lib64
+  - /share/aclocal
+  - /share/bakefile
+  - '*.la'
+  - '*.a'
+
+modules:
+  - shared-modules/glu/glu-9.json
+  - shared-modules/glew/glew.json
+  - shared-modules/SDL2/SDL2-with-libdecor.json
```
Comment (bbhtt) [inline com.vba_m.visualboyadvance-m.yaml:31]:
```suggestion
```

This is in 23.08

File: com.vba_m.visualboyadvance-m.yaml
Comment (bbhtt) [inline com.vba_m.visualboyadvance-m.yaml:1]: Delete this and rename the other one

**Comment by bbhtt** [comment]
bot, build com.vba_m.visualboyadvance-m

**Comment by ZachBacon** [comment]
I had a fix in place for the metainfo xml file, but I had failed to place it in the correct yaml. That's been fixed, so it'll have to go through another check. I'm sorry for that.

File: metainfo.patch [hunk 6]
```diff
@@ -0,0 +1,48 @@
...
++++ b/src/wx/visualboyadvance-m.metainfo.xml
+@@ -1,8 +1,9 @@
+ <?xml version="1.0" encoding="UTF-8"?>
+-<component type="desktop">
+- <id>visualboyadvance-m.desktop</id>
++<application>
++ <id type="desktop">visualboyadvance-m.desktop</id>
+  <metadata_license>CC-BY-SA-3.0</metadata_license>
+  <project_license>GPL-2.0</project_license>
++ <content_rating type="oars-1.0"/>
```
Comment (bbhtt) [inline metainfo.patch:26]:
```suggestion
+ <content_rating type="oars-1.1"/>
```

File: metainfo.patch [hunk 7]
```diff
@@ -0,0 +1,48 @@
...
+--- a/src/wx/visualboyadvance-m.metainfo.xml
++++ b/src/wx/visualboyadvance-m.metainfo.xml
+@@ -1,8 +1,9 @@
+ <?xml version="1.0" encoding="UTF-8"?>
+-<component type="desktop">
+- <id>visualboyadvance-m.desktop</id>
++<application>
++ <id type="desktop">visualboyadvance-m.desktop</id>
+  <metadata_license>CC-BY-SA-3.0</metadata_license>
+  <project_license>GPL-2.0</project_license>
```
Comment (bbhtt) [inline metainfo.patch:25]: Add `developer_name` too

File: metainfo.patch [hunk 8]
```diff
@@ -0,0 +1,51 @@
...
++++ b/src/wx/visualboyadvance-m.metainfo.xml
+@@ -1,8 +1,12 @@
+ <?xml version="1.0" encoding="UTF-8"?>
+-<component type="desktop">
+- <id>visualboyadvance-m.desktop</id>
++<application>
++ <id type="desktop">visualboyadvance-m.desktop</id>
++ <developer id="vba-m.com">
++   <name>VisualBoyAdvance-M</name>
++ </developer>
```
Comment (bbhtt) [inline metainfo.patch:26]: developer_id is not supported by Flathub I said developer_name

**Comment by bbhtt** [comment]
bot, build com.vba_m.visualboyadvance-m

**Comment by ZachBacon** [comment]
I already see one fail. Apparently it's not applying a patch properly that I made with git format-patch in vba-m's git master. I'll deal with this in the morning. It's already late for me.

**Comment by ZachBacon** [comment]
> I already see one fail. Apparently it's not applying a patch properly that I made with git format-patch in vba-m's git master. I'll deal with this in the morning. It's already late for me.

I actually know what was wrong. Foolish me decided that I'd try to apply a fix that I had done upstream against the tarball that didn't have the recent change. I'll do that up now.

**Comment by bbhtt** [comment]
You removed the screenshot/OARS from the patch

**Comment by ZachBacon** [comment]
> You removed the screenshot/OARS from the patch

Nope I didn't. Everything I did was pushed upstream. But because those changes aren't in a new release of vba-m yet. It wasn't in the tarball. The patch was trying to apply the changes in a file that didn't have those new additions, thus failing. I've fixed that, retesting locally. revalidating locally. It should pass this time once I push

**Comment by ZachBacon** [comment]
Regardless of which though. I'm catching some sleep. I'll deal with anything in the morning.

File: com.vba_m.visualboyadvance-m.yaml [hunk 9]
```diff
@@ -0,0 +1,77 @@
...
+  - --device=all
+  - --share=network
+  - --share=ipc
+  - --socket=fallback-x11
+  - --socket=wayland
+  - --socket=pulseaudio
+  - --filesystem=host:ro
+
+cleanup:
+  - /lib/cmake
```
Comment (bbhtt) [inline com.vba_m.visualboyadvance-m.yaml:19]: Add cleanups:

```
- /pkgconfig
- /share/doc
- /share/man
```

**Comment by bbhtt** [comment]
bot, build com.vba_m.visualboyadvance-m
